lib/src/presentation/tcp_scanner_task.dart

```diff
@@ -0,0 +1,106 @@
+import 'dart:math';
+
+import 'tcp_scanner_task_report.dart';
+import '../domain/entities/report.dart';
+import '../domain/interactor/scanner_isolate_interactor.dart';
+import '../presentation/tcp_scanner_task_exception.dart';
+
+class TcpScannerTask {
+  final String host;
+  late final List<int> ports;
+  final Duration socketTimeout;
+  final bool shuffle;
+  late final int parallelism;
+  bool _isRunning = false;
+  final List<ScannerIsolateInteractor> _scanners = [];
+
+  bool get isRunning => _isRunning;
+
+  TcpScannerTask(this.host, List<int> ports,
+      {this.socketTimeout = const Duration(seconds: 1), this.shuffle = false, int parallelism = 4}) {
+    //  Copy ports list and shuffle them if it's needed
+    var portsList = ports.toSet().toList();
+    if (shuffle) portsList.shuffle();
+    this.ports = portsList;
+    // Calculate number of isolates. The number of isolates can't be more than the number of ports.
+    this.parallelism = min(parallelism, ports.length);
+  }
+
+  /// Start scanning task
+  Future<TcpScannerTaskReport> start() async {
+    if (_isRunning) {
+      throw TcpScannerTaskException('Scanning is in progress');
+    }
+    _isRunning = true;
+    // Split the list of ports into sublists
+    var portsPerScanner = (ports.length / parallelism).ceil();
+    var sublistFirstIndex = 0;
+    var sublistLastIndex = 0;
+    while (sublistFirstIndex < ports.length) {
+      sublistLastIndex = min(sublistFirstIndex + portsPerScanner, ports.length);
+      var portsSublist = ports.sublist(sublistFirstIndex, sublistLastIndex);
+      _scanners.add(ScannerIsolateInteractor(host, portsSublist,
+          parallelism: parallelism, shuffle: shuffle, socketTimeout: socketTimeout));
+      sublistFirstIndex = sublistLastIndex;
+    }
+    // Start scan for each sublist
+    var reports = await Future.wait(_scanners.map((scanner) => scanner.scan()));
+    // Collect results to the resulting report
+    var scanReport = Report(host, ports, status: ReportStatus.finished);
+    reports.forEach((report) {
+      scanReport.addOpen(ports: report.openPorts);
+      scanReport.addClosed(ports: report.closedPorts);
+    });
+    _isRunning = false;
+    return _reportToTcpScannerTaskReport(scanReport);
+  }
+
+  /// Request scan report report
+  Future<TcpScannerTaskReport> get report async {
+    var scanReport = Report(host, ports);
+    var reports = await Future.wait(_scanners.map((scanner) => scanner.report));
```

Comment:
@shpak86 
Hi, when I run with parallelism 100 for all ports (65535 ports), and I get the report every single second.
I got this issue.

`[VERBOSE-2:ui_dart_state.cc(199)] Unhandled Exception: Bad state: Stream has already been listened to.
#0      _StreamController._subscribe (dart:async/stream_controller.dart:635:7)
#1      _ControllerStream._createSubscription (dart:async/stream_controller.dart:786:19)
#2      _StreamImpl.listen (dart:async/stream_impl.dart:473:9)
#3      Stream.first (dart:async/stream.dart:1211:14)
#4      ScannerIsolate.report (package:tcp_scanner/src/data/scanner_isolate.dart:55:49)
#5      ScannerIsolateInteractor.report (package:tcp_scanner/src/domain/interactor/scanner_isolate_interactor.dart:30:48)
#6      TcpScannerTask.report.<anonymous closure> (package:tcp_scanner/src/presentation/tcp_scanner_task.dart:61:72)
#7      MappedListIterable.elementAt (dart:_internal/iterable.dart:412:31)
#8      ListIterator.moveNext (dart:_internal/iterable.dart:341:26)
#9      Future.wait (dart:async/future.dart:393:26)
#10     TcpScannerTask.report (package:tcp_scanner/src/presentation/tcp_scanner_task.dart:61:32)
#11     <…>`

Reply:
Hi!
Don't scan port 0. If I'm not mistaken you can use it only as a source port in some reasons. 
It's not related with your problem, but decrease parallelism. You have no benefits due so many isolates. I bet if you used parallelism 30 or 50 you would have the same scan performance.

Reply:
@shpak86 thanks for your advice!

`Unhandled Exception: Bad state: Stream has already been listened to.`
Could you take a look at this issue?

Reply:
It was a mistake to throw the exception during the scan. I fixed it. Thank you!

Reply:
@shpak86 great work!!!
I used the latest version (2.0.4). It does not show the error.

I've just seen a new issues. When I run a high load (65535 ports - 50 isolates). You can see in the log, there's a long delay between the **first report** (firing at 3rd second) and the **second report** (firing at 10th second). After that, they are fired normally (every 2 seconds). I run a lot of tests and get the same phenomenon.
I think it's related with the previous issue `Unhandled Exception: Bad state: Stream has already been listened to.`

```
flutter: Host google.com scan progress 0.0%
Scanned ports:  8
Open ports:     []
Status: TcpScannerTaskReportStatus.progress
Elapsed:        **0:00:03.509615**
flutter: Host google.com scan progress 1.7%
Scanned ports:  1125
Open ports:     []
Status: TcpScannerTaskReportStatus.progress
Elapsed:        **0:00:10.683890**
flutter: Host google.com scan progress 2.5%
Scanned ports:  1671
Open ports:     [80]
Status: TcpScannerTaskReportStatus.progress
Elapsed:        **0:00:12.035338**
flutter: Host google.com scan progress 4.1%
Scanned ports:  2671
Open ports:     [80]
Status: TcpScannerTaskReportStatus.progress
Elapsed:        0:00:14.082610
```

Reply:
Hello!
I had similar behaviour during the testing. The reason is in the blocking event loop by isolates creation operations.
1. It takes long time to create 50 - 100 isolates. I added debug output to the isolate initializaton and found out that it takes about 4 seconds to create 100 isolates on my MacBook. It looks like dart creates groups of isolates by 16 pieces in each group and executes this groups. There is a long delay between groups execution, so I guess it's reasonable to use parallelism that is multiple of 16. Here is an example:
```
19:06:39.741823   <- isolate 1
19:06:39.744410
19:06:39.747574
19:06:39.749220
...
19:06:39.769956
19:06:39.777545
19:06:39.781630  <- isolate 16
19:06:40.018477  <- +0.237 second
19:06:40.020316  
...
19:06:43.836450
19:06:43.848669  <- isolate 100
```
2. Dart uses an event loop. This means that long queue of long tasks like isolates can cause delays. Your first report delays because event loop is blocked by bunch of long-term isolates initialisation operations. In your case it looks like dart creates isolated even after the first `print` operation and this is a reason why you have a delay between first two prints. The timer which calls `print` just blocked by isolates creation. 

Unfortunately, I don't have a solution for this issue right now. I can only advice you to use less `parallelism` values to avoid blocking event loop by high number of isolates creation operations.

Reply:
@shpak86 
I got it. Thank you very for support quickly as well as great advices!

The last one I would like to let you know.
When I try to test on real device (iPhone) with a small number of ports (260 ports) and 10 parallelism.
After scanning 4-5 times, I got the `Out of memory` issue.
Do you have any ideas to fix it?

```
Exhausted heap space, trying to allocate 60448 bytes.
../../third_party/dart/runtime/vm/object.cc: 2625: error: Out of memory.
version=2.13.4 (stable) (Wed Jun 23 13:08:41 2021 +0200) on "ios_arm64"
pid=14888, thread=698147, isolate_group=_scan(0x11fbbde00), isolate=_scan(0x11fa99e00)
isolate_instructions=10288c4e0, vm_instructions=10288c4e0
Exhausted heap space, trying to allocate 58288 bytes.
../../third_party/dart/runtime/vm/object.cc: 2625: error: Out of memory.
Aborting re-entrant request for stack trace.
* thread #29, name = 'DartWorker', stop reason = signal SIGABRT
    frame #0: 0x00000001ae4fe7b0 libsystem_kernel.dylib`__pthread_kill + 8
libsystem_kernel.dylib`__pthread_kill:
->  0x1ae4fe7b0 <+8>:  b.lo   0x1ae4fe7cc               ; <+36>
    0x1ae4fe7b4 <+12>: stp    x29, x30, [sp, #-0x10]!
    0x1ae4fe7b8 <+16>: mov    x29, sp
    0x1ae4fe7bc <+20>: bl     0x1ae4db1fc               ; cerror_nocancel
Target 0: (Runner) stopped.
```

Reply:
Thank you for the info! This is a memory leak. I fixed it in version 2.0.5.

Reply:
@shpak86 
Awesome!!!
You are so quick, I've just tried to fix it :)
https://github.com/phan-dev/tcp_scanner/commit/941a94b1e6d9f956eaed6ba251a11361ffe71097